File: web/ViewModels/BookmarkViewModel.cs
using web.Models;
using web.Services;

namespace web.ViewModels
{
    public enum BookmarkState { Idle, Loading, Success, Error }
    public class BookmarkViewModel
    {
        private readonly ApiService _apiService;

        public BookmarkState State { get; private set; } = BookmarkState.Idle;
        public List<BookmarkModel> Bookmarks { get; private set; } = new();
        public string ErrorMessage { get; private set; } = string.Empty;
        public bool IsEmpty => Bookmarks.Count == 0 && State == BookmarkState.Success;

        public event Action? OnChange;

        public BookmarkViewModel(ApiService apiService)
        {
            _apiService = apiService;
        }

        public async Task FetchBookmarksAsync()
        {
            SetState(BookmarkState.Loading);

            try
            {
                Bookmarks = await _apiService.GetBookmarksAsync();
                SetState(BookmarkState.Success);
            } catch(Exception ex)
            {
                ErrorMessage = ex.Message;
                SetState(BookmarkState.Error);
            }
        }
        public async Task RemoveBookmarkAsync(string articleId)
        {
            var removed = Bookmarks.FirstOrDefault(b => b.Article.Id == articleId);
            if (removed == null) return;

            Bookmarks = Bookmarks.Where(b => b.Article.Id != articleId).ToList();
            NotifyStateChanged();

            try
            {
                await _apiService.RemoveBookmarkAsync(articleId);
            }
            catch
            {
                Bookmarks = new List<BookmarkModel>(Bookmarks) { removed };
                NotifyStateChanged();
            }
        }
        public async Task AddBookmarkAsync(string articleId)
        {
            try
            {
                await _apiService.AddBookmarkAsync(articleId);
                await FetchBookmarksAsync();
            } catch(Exception ex)
            {
                ErrorMessage = ex.Message;
                NotifyStateChanged();
            }
        }
        public bool IsBookmarked(string articleId)
        {
            return Bookmarks.Any(b => b.Article.Id == articleId);
        }

        private void SetState(BookmarkState state)
        {
            State = state;
            NotifyStateChanged();
        }
        private void NotifyStateChanged() => OnChange?.Invoke();
    }
}

File: web/js/viewmodels/BookmarkViewModel.js
import { getBookmarks, addBookmark, removeBookmark } from '../services/apiService';
import { Bookmark } from '../models/Bookmark';

export class BookmarkViewModel {
    #bookmarks = [];
    #state = 'idle';
    #errorMessage = '';
    #listeners = new  Set();

    get bookmarks() {return this.#bookmarks; }
    get state() {return this.#state; }
    get errorMessage() { return this.#errorMessage }
    get isEmpty() { return this.#bookmarks.length === 0 && this.#state === 'success'; }
    get count() { return this.#bookmarks.length; }

    subscribe(fn) {
        this.#listeners.add(fn);
        return () => this.#listeners.delete(fn);
    }

    #notify() {
        this.#listeners.forEach((fn) => fn(this));
    }

    #setState(state) {
        this.#state = state;
        this.#notify();
    }

    async fetchBookmarks() {
        this.#setState('loading');
        try {
            const data = await getBookmarks();
            this.#bookmarks = Bookmark.fromList(data);
            this.#setState('success');
        } catch(err) {
            this.#errorMessage = err.message;
            this.#setState('error');
        }
    }

    async addBookmark(articleId) {
        try{
            await addBookmark(articleId);
            await this.fetchBookmarks();
        } catch(err) {
            this.#errorMessage = err.message;
            this.#notify();
        }
    }

    async removeBookmark(articleId) {
        const removed = this.#bookmarks.find((b) => b.article.id === articleId);
        if(!removed) return;

        this.#bookmarks = this.#bookmarks.filter((b) => b.article.id !== articleId);
        this.#notify();

        try {
            await removeBookmark(articleId);
        } catch{
            this.#bookmarks = [...this.#bookmarks, removed];
            this.#notify();
        }
    }

    isBookmarked(articleId) {
        return this.#bookmarks.some((b) => b.article.id === articleId);
    }
}